### PR TITLE
Set node engine to 7.6 and refactor to use ES6 class and lambda syntax

### PR DESCRIPTION
**Pull request by spoon16**
I threw this together while I was waiting for a new environment to come up in AWS because I heard you say you were interesting in people helping you move your repositories to more modern JS syntax/patterns. This does break backwards compatibility with some old versions of node. You could probably fix that with webpack/babel if maintaining that compatibility is important.

Set minimum node version in package.json to support ES6 features. Refactored `avrgirlUsbTinyIsp` to be an ES6 class. Refactored most the the callbacks to use lambda syntax. Refactored some string concatenation to use template literals.

This diff is a bit gross because of the refactor to ES6 class syntax. Feel free to reject if that's frustrating.

### CODE REVIEW

File: avrgirl-usbtinyisp.js [hunk 1]
```diff
@@ -1,432 +1,414 @@
...
+const EventEmitter = require('events').EventEmitter;
+const usbtinyisp = require('usbtinyisp');
+const bufferEqual = require('buffer-equal');
+const async = require('async');
+const programmers = require('./lib/programmers');
+const intelhex = require('intel-hex');
+const fs = require('fs');
+const chips = require('avrgirl-chips-json');
+
+class avrgirlUsbTinyIsp extends EventEmitter {
```
Comment (spoon16) [inline avrgirl-usbtinyisp.js]: switched to ES6 class syntax here.

File: avrgirl-usbtinyisp.js [hunk 2]
```diff
@@ -1,432 +1,414 @@
...
+    // fix this pls self, it's very unattractive
+    // do an error check too
+    var p = self.options.programmer ? programmers[self.options.programmer] : null;
+    self.options.pid = p ? p.pid : 3231;
+    self.options.vid = p ? p.vid : 6017;
+
+    // create new instance of usbtiny isp as programmer instance
+    self.programmer = new usbtinyisp(self.options);
+
+    self.programmer.open(error => {
```
Comment (spoon16) [inline avrgirl-usbtinyisp.js]: lambda syntax here

File: avrgirl-usbtinyisp.js [hunk 3]
```diff
@@ -1,432 +1,414 @@
...
+        if (e.code === 'ENOENT') {
+          return callback(new Error(`could not write ${memType}: please supply a valid path to a hex file.`));
+        }
         return callback(e);
       }
+
+      data = intelhex.parse(readFile).data;
+
+    } else {
+      return callback(new Error(`could not write ${memType}: please supply either a hex buffer or a valid path to a hex file.`));
```
Comment (spoon16) [inline avrgirl-usbtinyisp.js]: template literals here

File: package.json [hunk 4]
```diff
@@ -3,6 +3,9 @@
   "version": "1.0.3",
   "description": "avrgirl wrapper for usbtinyisp programmer interactions",
   "main": "avrgirl-usbtinyisp.js",
+  "engines": {
+    "node": ">=7.6"
```
Comment (spoon16) [inline package.json]: set minimum node engine version here

File: avrgirl-usbtinyisp.js [hunk 5]
```diff
@@ -1,432 +1,414 @@
...
+  constructor(options) {
+    super();
+
+    var self = this;
+
+    // set up noisy or quiet
+    self.debug = options.debug ? console.log.bind(console) : () => {};
+
+    // most people won't need this level of debug output
+    self.hackerLog = options.hackerMode ? console.log.bind(console) : () => {};
```
Comment (spoon16) [inline avrgirl-usbtinyisp.js]: bind console.log here as is done on l:24

File: avrgirl-usbtinyisp.js [hunk 6]
```diff
@@ -1,432 +1,414 @@
...
-avrgirlUsbTinyIsp.prototype.enterProgrammingMode = function (callback) {
-  var self = this;
-  var cmd = new Buffer(this.options.chip.pgmEnable);
-
-  this.setSCK(this.options.sck, function(error) {
-    if (error) { return callback(error) }
-    // is this timeout necessary?
-    setTimeout(function() {
-      self.programmer.spi(cmd, function (error) {
+    self.setSCK(self.options.sck, error => {
```
Comment (spoon16) [inline avrgirl-usbtinyisp.js]: There were a couple places where `var self = this;` pattern is used, but `self` was not referenced consistently in the body of the function. Right here I changed `this` to `self` to be more consistent.

File: avrgirl-usbtinyisp.js [hunk 7]
```diff
@@ -1,432 +1,414 @@
...
- * @param {function} callback - function to run upon completion/error
- */
-avrgirlUsbTinyIsp.prototype.setSCK = function (rate, callback) {
-  // preparing for next version is usbtinyisp to be published allowing for custom clock rate
-  if (!callback) {
-    var callback = rate;
+      }
+      // is this timeout necessary?
+      setTimeout(() =>
+        self.programmer.spi(cmd, error => callback(error)), 50);
```
Comment (spoon16) [inline avrgirl-usbtinyisp.js]: lambda syntax

File: avrgirl-usbtinyisp.js [hunk 8]
```diff
@@ -1,432 +1,414 @@
...
+   */
+  setSCK(rate, callback) {
+    // preparing for next version is usbtinyisp to be published allowing for custom clock rate
+    if (!callback) {
+      callback = rate;
+    }
+
+    // error checking for rate being a number
+    if ((typeof rate !== 'number') || (rate < C.SCK_MIN) || (rate > C.SCK_MAX)) {
+      return callback(new Error(`Could not set SCK: rate should be a number between ${C.SCK_MIN} and ${C.SCK_MAX}.`));
```
Comment (spoon16) [inline avrgirl-usbtinyisp.js]: template string

File: avrgirl-usbtinyisp.js [hunk 9]
```diff
@@ -1,432 +1,414 @@
...
+    var signature = this.options.chip.signature;
+    var cmd = new Buffer(signature.read);
+    var sigLen = signature.size;
+    var set = 0;
+    var sigPos = 3;
+
+    var self = this;
+
+    // recursive function, according to length of the signature requested
+    const getSigByte = () => {
```
Comment (spoon16) [inline avrgirl-usbtinyisp.js]: lambda syntax

**Comment by noopkat** [comment]
hi @spoon16!

Sorry this took an extra week to get to. I went ahead and installed babel and am transpiling for compatibility with Node v4/5.

Thanks for an excellent PR ✨  I'm merging and publishing this along with the babel work.

bitmoji

**Comment by spoon16** [comment]
I did not expect you to merge. Thanks!

**Pull request merged**